SYNTHESIZER: Handle the case of violation in loop guards

qinheping opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

In current loop-invariant synthesizer, we synthesize two kinds of clauses separately.
1. `in_clause` that should hold when we enter the loop body---when the loop guard holds, that is ```loop_guard -> in_clause```.
2. `pos_clause` that should hold when we leave the loop---when the loop guard doesn't hold, that is ```!loop_guard -> pos_clause```.

We synthesize `in_clause` when the violations happen in the loop body. We synthesize `pos_clause` when the violation happen after the loop.

This PR handles the case that the violations happen in the loop guard ([loop_contracts_synthesis_06/main.c](https://github.com/diffblue/cbmc/compare/develop...qinheping:goto-synthesizer-violation-location?expand=1#diff-8a05899ddc66b61918f2e6180bfc08f09f895f6a78066c65086a5e7c470a880a) as an example). In the case, the new clause we synthesize should hold for both of the cases that loop guard holds and loop guards doesn't hold.

We introduce a new variable `violation_location` to indicate whether the violation happen in the loop, after the loop, or in the loop guard, and add the new synthesized clause as `in_clause`, `pos_clause`, or both, respectively.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

feliperodri commented on regression/goto-synthesizer/loop_contracts_synthesis_06/main.c:
```diff
@@ -0,0 +1,18 @@
+#define SIZE 80
```
Include `stdlib`.

feliperodri commented on regression/goto-synthesizer/loop_contracts_synthesis_06/main.c:
```diff
@@ -0,0 +1,18 @@
+  __CPROVER_assume(len <= SIZE);
+  __CPROVER_assume(len >= 8);
+  char *array = malloc(len);
+  __CPROVER_assume(array != 0);
```
This assume is unnecessary. Without malloc flags, this malloc will always return a valid allocation.

qinheping replied:
Yes, it is unnecessary now. But the malloc flags may become default flag in the future.
https://github.com/diffblue/cbmc/pull/7448#discussion_r1053454223

feliperodri commented on regression/goto-synthesizer/loop_contracts_synthesis_06/test.desc:
```diff
@@ -0,0 +1,13 @@
+--
+--
+This test is a variation of strlen. It checks that we can synthesize
+loop invariant for checks happen in the loop guard.
```
```suggestion
Checks whether CBMC synthesizes loop invariants for checks in the loop guard.
```

feliperodri commented on regression/goto-synthesizer/loop_contracts_synthesis_06/test_dump.desc:
```diff
@@ -0,0 +1,9 @@
+loop 1 invariant.*\_\_CPROVER\_POINTER\_OFFSET
+assigns i
+--
+This test case checks if synthesized contracts are correctly dumped.
```
```suggestion
Checks if synthesized contracts are dumped correctly.
```

feliperodri commented on regression/goto-synthesizer/loop_contracts_synthesis_07/main.c:
```diff
@@ -0,0 +1,18 @@
+#define SIZE 80
```
Includes `stdlib`.

feliperodri commented on regression/goto-synthesizer/loop_contracts_synthesis_07/test.desc:
```diff
@@ -0,0 +1,14 @@
+--
+--
+This test checks that we can synthesize loop invariant for violation happen
+after the loop body.
```
```suggestion
Checks whether CBMC synthesizes loop invariants for checks located after the loop body.
```

feliperodri commented on regression/goto-synthesizer/loop_contracts_synthesis_07/test_dump.desc:
```diff
@@ -0,0 +1,9 @@
+loop 1 assigns i
+loop 1 invariant.*\_\_CPROVER\_POINTER\_OFFSET
+--
+This test case checks if synthesized contracts are correctly dumped.
```
```suggestion
Checks whether synthesized contracts are dumped correctly.
```

feliperodri commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -211,6 +211,62 @@ std::list<loop_idt> cegis_verifiert::get_cause_loop_id(
+      it != function.body.instructions.end();
+      ++it)
+  {
+    // Record the location number of the begin of a transformed loop.
```
```suggestion
    // Record the location number of the beginning of a transformed loop.
```

feliperodri commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -211,6 +211,62 @@ std::list<loop_idt> cegis_verifiert::get_cause_loop_id(
+  const goto_functiont &function,
+  unsigned location_number_of_target)
+{
+  // The transformed loop condition are instructions from
```
```suggestion
  // The transformed loop condition is a set of instructions from
```

feliperodri commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -211,6 +211,62 @@ std::list<loop_idt> cegis_verifiert::get_cause_loop_id(
+  // loop havocing instructions
+  // to
+  // if(!guard) GOTO EXIT
+  unsigned location_number_of_havocing = 0;
+  for(goto_programt::const_targett it = function.body.instructions.begin();
+      it != function.body.instructions.end();
+      ++it)
+  {
```
```suggestion
  for(auto it = function.body.instructions.begin();
      it != function.body.instructions.end();
      ++it)
  {
    unsigned location_number_of_havocing = 0;
```

qinheping replied:
`location_number_of_havocing` is assigned and used in different loop iterations. It should not be a loop-local here.

feliperodri commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -404,14 +404,33 @@ invariant_mapt enumerative_loop_contracts_synthesizert::synthesize_all()
+        return_cex->violation_location ==
+        cext::violation_locationt::in_condition)
+      {
+        // When the violation is happen in the loop guard, the new clause
```
```suggestion
        // When the violation happens in the loop guard, the new clause
```

feliperodri commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -404,14 +404,33 @@ invariant_mapt enumerative_loop_contracts_synthesizert::synthesize_all()
+      else if(
+        return_cex->violation_location == cext::violation_locationt::in_loop)
       {
+        // When the violation is happen in the loop body, the new clause
```
```suggestion
        // When the violation happens in the loop body, the new clause
```

feliperodri commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -404,14 +404,33 @@ invariant_mapt enumerative_loop_contracts_synthesizert::synthesize_all()
       else
       {
-        // violation happens post-loop.
+        // When the violation is happen after the loop body, the new clause
```
```suggestion
        // When the violation happens after the loop body, the new clause
```

feliperodri commented on src/goto-synthesizer/cegis_verifier.h:
```diff
@@ -61,9 +68,8 @@ class cext
-  // true if the violation happens in the cause loop
-  // false if the violation happens after the cause loop
-  bool is_violation_in_loop = true;
+  // Location where the violation happen.
```
```suggestion
  // Location where the violation happens
```

feliperodri commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -211,6 +211,62 @@ std::list<loop_idt> cegis_verifiert::get_cause_loop_id(
+  // to
+  // if(!guard) GOTO EXIT
+  unsigned location_number_of_havocing = 0;
+  for(goto_programt::const_targett it = function.body.instructions.begin();
```
```suggestion
  for(auto it = function.body.instructions.begin();
```

qinheping committed:
Handle the case of violation in loop guards in loop-invariant synthesis

In current loop-invariant synthesizer, we synthesize two kinds of clauses separately.
1. `in_clause` that should hold when we enter the loop body---when the loop guard holds, that is ```loop_guard -> in_clause```.
2. `pos_clause` that should hold when we leave the loop---when the loop guard doesn't hold, that is ```!loop_guard -> pos_clause```.

We synthesize `in_clause` when the violations happen in the loop body. We synthesize `pos_clause` when the violation happen after the loop.

This PR handles the case that the violations happen in the loop guard. In the case, the new clause we synthesize should hold for both of the cases that loop guard holds and loop guards doesn't hold.

We introduce a new variable `violation_location` to indicate whether the violation happen in the loop, after the loop, or in the loop guard, and add the new synthesized clause as `in_clause`, `pos_clause`, or both, respectively.
